subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalCompilerFactory.java

```diff
@@ -98,7 +98,11 @@ public PreviousCompilationStore getPreviousCompilationStore() {
             @Override
             public WorkResult execute(T spec) {
                 spec.setSourceFiles(sourceFiles);
-                return cleaningJavaCompiler.execute(spec);
+                WorkResult result = cleaningJavaCompiler.execute(spec);
+                if (result instanceof GroovyCompileResult) {
```

Comment:
I think it is very strange that the `CompilerFactory` knows about `GroovyCompileResult`...

Reply:
Yeah.... do you have better ideas?

Reply:
We already have an `IncrementalCompilerDecorator`. We could have our own decorator for Groovy, which then returns `GroovyCompileResult` (or `IncrementalCompilationResult` to notify that the compilation was performed incrementally. In all other cases, a "regular" `WorkResult` can be returned. I guess `RecompilationNotNecessary` is also an `IncrementalCompilationResult`.

Reply:
And extra `IncrementalCompilerDecorator` for Groovy would be a little heavy, I added a method to `RecompilationSpecProvider` to decorate the result.